### PR TITLE
bsp: Add ESP32-LyraT Board Support Package

### PR DESCRIPTION
**Pull request by espzav**
# Checklist for new Board Support package or Component

- [x] Component contains License
- [x] Component contains README.md
- [x] Project [README.md](../README.md) updated
- [x] Component contains idf_component.yml file with `url` field defined
- [x] Component was added to CI [upload job](https://github.com/espressif/esp-bsp/blob/master/.github/workflows/upload_component.yml#L17)
- [x] New files were added to CI build job
- [x] New BSP definitions added to [bsp_ext.py](../examples/bsp_ext.py)
- [ ] _Optional:_ Component contains unit tests
- [ ] CI passing

# Change description
Added new BSP: ESP32-LyraT. There is still noise in audio, when compiled in IDF5.1

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -15,6 +15,7 @@ Board support packages for development boards using Espressif's SoCs, written in
 | [ESP32-S3-EYE](esp32_s3_eye) | ESP32-S3 | LCD display, camera, uSD card slot, microphone and accelerometer | <img src="docu/pics/s3-eye.webp" width="150"> |
 | [ESP32-S3-LCD-EV-BOARD](esp32_s3_lcd_ev_board) | ESP32-S3 | LCD display with touch, audio codec + power amplifier | <img src="docu/pics/esp32-s3-lcd-ev-board_800x480.png" width="150"> |
 | [ESP32-S3-Korvo-2](esp32_s3_korvo_2) | ESP32-S3 | LCD display, camera, uSD card slot, microphone, audio codec + power amplifier | <img src="docu/pics/korvo-2.png" width="150"> |
+| [ESP32-LyraT](esp32_lyrat) | ESP32 | uSD card slot, microphone, audio codec + power amplifier | <img src="docu/pics/lyrat.png" width="150"> |
```
Comment (tore-espressif) [inline README.md:18]: Could we add a picture without the TEM watermark?

File: esp32_lyrat/Kconfig [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+            help
+                Supported max files for SPIFFS in the Virtual File System.
+    endmenu
+
+    config BSP_I2S_NUM
+        int "I2S peripheral index"
+        default 1
+        range 0 1
+        help
+            ESP32S3 has two I2S peripherals, pick the one you want to use.
```
Comment (tore-espressif) [inline esp32_lyrat/Kconfig:75]: Nitpick:
```suggestion
            ESP32 has two I2S peripherals, pick the one you want to use.
```

File: esp32_lyrat/idf_component.yml [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+dependencies:
+  idf: ">=4.4"
+
+  button:
+    version: "^2"
+    public: true
+
+  esp_codec_dev:
+    version: "^1"
```
Comment (tore-espressif) [inline esp32_lyrat/idf_component.yml:16]: Let's enforce at least 1.0.2, it contains important bugfixes

File: esp32_lyrat/idf_component.yml [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+url: https://github.com/espressif/esp-bsp/tree/master/esp32_lyrat
+
+targets:
+  - esp32
+
+dependencies:
+  idf: ">=4.4"
+
+  button:
+    version: "^2"
```
Comment (tore-espressif) [inline esp32_lyrat/idf_component.yml:12]: Custom buttons are supported from v2.5

File: esp32_lyrat/include/bsp/esp32_lyrat.h [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+ *
+ * I2C interface
+ *
+ * There are multiple devices connected to I2C peripheral:
+ *  - Codec ES8388 (configuration only)
+ *
+ * After initialization of I2C, use BSP_I2C_NUM macro when creating I2C devices drivers ie.:
+ * \code{.c}
+ * es8311_handle_t es8311_dev = es8311_create(BSP_I2C_NUM, ES8311_ADDRESS_0);
+ * \endcode
```
Comment (tore-espressif) [inline esp32_lyrat/include/bsp/esp32_lyrat.h:220]: This is not recommended way of interfacing the codec. Can be removed

File: esp32_lyrat/include/bsp/esp32_lyrat.h [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+ *
+ * I2S audio interface
+ *
+ * Devices connected to the I2S peripheral:
+ *  - Codec ES8388 for output (playback) path
+ *
+ * After initialization of I2S with bsp_audio_init(), use standard I2S drive API for reading/writing to I2S stream:
+ * \code{.c}
+ * i2s_write_channel(tx_handle, wav_bytes, wav_bytes_len, &i2s_bytes_written, pdMS_TO_TICKS(500));
+ * \endcode
```
Comment (tore-espressif) [inline esp32_lyrat/include/bsp/esp32_lyrat.h:90]: Please update the docs for esp_codec_dev, Something like this:
```
For speaker initialization use bsp_audio_codec_speaker_init() which is inside initialize I2S with bsp_audio_init().
 * For microphone initialization use bsp_audio_codec_microphone_init() which is inside initialize I2S with bsp_audio_init().
 * After speaker or microphone initialization, use functions from esp_codec_dev for play/record audio.
 * Example audio play:
 * \code{.c}
 * esp_codec_dev_set_out_vol(spk_codec_dev, DEFAULT_VOLUME);
 * esp_codec_dev_open(spk_codec_dev, &fs);
 * esp_codec_dev_write(spk_codec_dev, wav_bytes, wav_bytes_len);
 * esp_codec_dev_close(spk_codec_dev);
 * \endcode
 ```

File: esp32_lyrat/include/bsp/esp32_lyrat.h [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+ * Buttons interface
+ *
+ * Example configuration:
+ * \code{.c}
+ * button_handle_t button[BSP_BUTTON_NUM];
+ * for (int i = 0; i < BSP_BUTTON_NUM; i++) {
+ *     button[i] = iot_button_create(&bsp_button_config[i]);
+ * }
+ * \endcode
+ **************************************************************************************************/
```
Comment (tore-espressif) [inline esp32_lyrat/include/bsp/esp32_lyrat.h:67]: Could we provide all-buttons-init function `esp_err_t bsp_iot_button_create(button_handle_t btn_array[], int *btn_cnt, int btn_array_size);` and update this comment

```
* Example configuration:
 * \code{.c}
 * button_handle_t btns[BSP_BUTTON_NUM];
 * bsp_iot_button_create(btns, NULL, BSP_BUTTON_NUM);
 * iot_button_register_cb(btns[0], ...
 * \endcode
```

**Review by tore-espressif — COMMENTED** [review]
First round of review, so you are not blocked. I'll continue ASAP

File: esp32_lyrat/esp32_lyrat_idf4.c [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+    /* Setup power amplifier pin */
+    const gpio_config_t io_conf = {
+        .intr_type = GPIO_INTR_DISABLE,
+        .mode = GPIO_MODE_OUTPUT,
+        .pin_bit_mask = BIT64(BSP_POWER_AMP_IO),
+        .pull_down_en = GPIO_PULLDOWN_DISABLE,
+        .pull_up_en = GPIO_PULLDOWN_DISABLE,
+    };
+    ESP_ERROR_CHECK(gpio_config(&io_conf));
```
Comment (tore-espressif) [inline esp32_lyrat/esp32_lyrat_idf4.c:49]: The PA pin should now be managed by esp_codec_dev. See https://github.com/espressif/esp-bsp/blob/31184767fc82edcd40ce05a7ed26b6390f94d30f/esp32_lyrat/esp32_lyrat.c#L188

So all API that init/deinit/control this pin should be removed

File: esp32_lyrat/esp32_lyrat_idf4.c [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+#include "esp_err.h"
+#include "bsp/esp32_lyrat.h"
+#include "bsp_err_check.h"
+#include "esp_codec_dev_defaults.h"
+
+static const audio_codec_data_if_t *i2s_data_if = NULL;  /* Codec data interface */
+
+esp_err_t bsp_audio_init(const i2s_config_t *i2s_config)
+{
```
Comment (tore-espressif) [inline esp32_lyrat/esp32_lyrat_idf4.c:15]: Could we check `i2s_data_if != NULL` in the beginning and return ESP_OK if the condition is true?

File: esp32_lyrat/esp32_lyrat.c [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+
+static const char *TAG = "LyraT";
+
+sdmmc_card_t *bsp_sdcard = NULL;    // Global uSD card handler
+
+static esp_err_t  bsp_touchpad_custom_init(void *param);
+static esp_err_t  bsp_touchpad_custom_deinit(void *param);
+static uint8_t bsp_touchpad_custom_get_key_value(void *param);
+
+const button_config_t bsp_button_config[BSP_BUTTON_NUM] = {
```
Comment (tore-espressif) [inline esp32_lyrat/esp32_lyrat.c:25]: nitpick: static

File: esp32_lyrat/esp32_lyrat.c [hunk 11]
```diff
@@ -0,0 +1,278 @@
...
+    {
+        .type = BUTTON_TYPE_CUSTOM,
+        .custom_button_config.button_custom_init = bsp_touchpad_custom_init,
+        .custom_button_config.button_custom_get_key_value = bsp_touchpad_custom_get_key_value,
+        .custom_button_config.button_custom_deinit = bsp_touchpad_custom_deinit,
+        .custom_button_config.priv = (void *)BSP_BUTTON_VOLUP,
+    }
+};
+
+esp_err_t bsp_i2c_init(void)
```
Comment (tore-espressif) [inline esp32_lyrat/esp32_lyrat.c:66]: To allow multiple calls to bsp_i2c_init
``` c
esp_err_t bsp_i2c_init(void)
{
    /* I2C was initialized before */
    if (i2c_initialized) {
        return ESP_OK;
    }

    const i2c_config_t i2c_conf = {
        .mode = I2C_MODE_MASTER,
        .sda_io_num = BSP_I2C_SDA,
        .sda_pullup_en = GPIO_PULLUP_DISABLE,
        .scl_io_num = BSP_I2C_SCL,
        .scl_pullup_en = GPIO_PULLUP_DISABLE,
        .master.clk_speed = CONFIG_BSP_I2C_CLK_SPEED_HZ
    };
    BSP_ERROR_CHECK_RETURN_ERR(i2c_param_config(BSP_I2C_NUM, &i2c_conf));
    BSP_ERROR_CHECK_RETURN_ERR(i2c_driver_install(BSP_I2C_NUM, i2c_conf.mode, 0, 0, 0));

    i2c_initialized = true;

    return ESP_OK;
}

esp_err_t bsp_i2c_deinit(void)
{
    BSP_ERROR_CHECK_RETURN_ERR(i2c_driver_delete(BSP_I2C_NUM));
    i2c_initialized = false;
    return ESP_OK;
}
```

**Review by tore-espressif — COMMENTED** [review]
Few more comments, but nothing major. LGTM

File: README.md (hunk 1 above)
Comment (espzav) [reply README.md:18]: Fixed. Sorry, I found bad image and I didn't see it on my screen before (maybe lot of sun light).

**Comment by espzav** [comment]
@tore-espressif Thank you for review. All fixed, but the recording still not working. I have two errors there and second one crash:
```
**E (19429) i2s_common: i2s_channel_disable(1030): the channel has not been enabled yet**
I (19429) I2S_IF: channel mode 0 bits:16/16 channel:2 mask:1
I (19449) I2S_IF: STD Mode 1 bits:16/16 channel:2 sample_rate:16000 mask:1
I (19459) Adev_Codec: Open codec device OK
I (19459) example: Recording start
**E (20459) i2s_common: i2s_channel_read(1164): The channel is not enabled**
```

**Comment by tore-espressif** [comment]
> @tore-espressif Thank you for review. All fixed, but the recording still not working. I have two errors there and second one crash:
> 
> ```
> E (19429) i2s_common: i2s_channel_disable(1030): the channel has not been enabled yet
> I (19429) I2S_IF: channel mode 0 bits:16/16 channel:2 mask:1
> I (19449) I2S_IF: STD Mode 1 bits:16/16 channel:2 sample_rate:16000 mask:1
> I (19459) Adev_Codec: Open codec device OK
> I (19459) example: Recording start
> E (20459) i2s_common: i2s_channel_read(1164): The channel is not enabled
> ```

This is almost sure related to https://github.com/espressif/esp-adf/issues/1024

**Comment by TempoTian** [comment]
@espzav  the error log is just a warning, for some codec need i2s clock before codec initialize, so i2s enable is not fully controlled in data_i2s.  data_i2s can't get i2s really status, so there are some log like `already disabled`

**Comment by espzav** [comment]
@tore-espressif please, is it ok? Can we continue with review and merge it after?

File: esp32_lyrat/include/bsp/esp32_lyrat.h (hunk 5 above)
Comment (tore-espressif) [reply esp32_lyrat/include/bsp/esp32_lyrat.h:174]: I mean that the whole section
```
 After initialization of I2C, use BSP_I2C_NUM macro when creating I2C devices drivers ie.:
 * \code{.c}
 * es8311_handle_t es8311_dev = es8311_create(BSP_I2C_NUM, ES8311_ADDRESS_0);
 * \endcode
 ```
can be removed, right?

**Review by tore-espressif — APPROVED** [review]
Only one nitpick, otherwise LGTM!

**Pull request merged**